### PR TITLE
Disallow Yoda Comparison

### PR DESCRIPTION
**Pull request by gmsantos**
When we started to discuss about Coding Standards, we already had rules that [goes far from PSR-2](https://github.com/leroy-merlin-br/kameleon/wiki/Extended-Coding-Style-Guide/a14db5c0c1f84df26a46930b68659e6f9e5aad22). 

We decided to adopt [Symfony Coding Standards](https://github.com/leroy-merlin-br/kameleon/wiki/Extended-Coding-Style-Guide/bf42ece9d8b14047186aa61282b2c90a134f36c7) (SCS) because it has lot of rules similar to our needs and it is already used by Symfony Core Team. However SCS has newer rules and we adopted too to keep it consistent.

One of these rules is Yoda Comparison.  At the time was a disputed votation with only one vote in diference. 

Now that our coding standard is more granular, we could reevaluate this rule and discuss if:

- We need to enforce it (Current behavior)
- We could disallow it (this PR)
- We ask for "vodka or coconut water" (neither enforce or disallow - remove this rule) 

Reference: https://github.com/slevomat/coding-standard#slevomatcodingstandardcontrolstructuresdisallowyodacomparisonrequireyodacomparison-

### CODE REVIEW

**Comment by vitorbari** [comment]
I always have to over think when using yoda style in comparisons such as `<`, `>=`

**Review by carusogabriel — DISMISSED** [review]
:-1: for this one, and one of the main reason is if we `misspelling` the equal operator `==`:

```php
if (!$this->isValid() = $foo) {
```

without Yoda Comparison, might take a while for the developer to figure it out. Besides that, with Yoda, as the `!$foo = $this->isValid()` is an invalid statment, would take seconds to catch it.

**Comment by alemferreira** [comment]
Can we not add a new rule that checks inside comparison structures has only one `=`? If it has, the CS will trigger. There is cases that we use only `=` in comparision structures like:
```
if ($foo = $bar->someFunction()) {}
``` 
In this particular case, the $foo is a attribution and the `if` will check if the variable exists, for those cases, we would need to refactor to:

```
$foo = $bar->someFunction();
if ($foo) {}
```

This way we could both drop Yoda comparision and mantain a readable conditional structures using `< and <=`

**Comment by gmsantos** [comment]
@carusogabriel is this misspelling really happen with us? 

This kind of situation could be detected by unit tests, as @brnbp stated in other PR. Since we encourage high unit test coverage, an untested `if` case would not happen.

As @vitorbari said, other disvantage is to flip `>` and `<=` operator.

And `!$foo = $this->isValid()` actually works:

https://3v4l.org/HMKd0

**Review by alemferreira — DISMISSED** [review]

**Comment by ravanscafi** [comment]
@aleferreiranogueira there are several cases where `=` is desirable:

```php
if (!$product = $repository->first()) {
    return false;
}
```

Honestly, for me, it makes no sense. The yoda rule prevents inexperienced developers from doing this kind of mistake. We easily spot this kinda of problem and I can't remember at least one time that we had this problem in the past years.
I'm ok with both decisions.

**Comment by alemferreira** [comment]
@ravanscafi fair enough, I'm ok with only disallowing the Yoda comparision, I presented that alternative only to make sure that we wouldn't have attributions when it should've been a comparision; However, as @gmsantos stated, that can be prevented by testing and this behaviour is desirable (and used, like you said) throughout our application

**Comment by gmsantos** [comment]
@aleferreiranogueira there is a rule that is recommend to use together with yoda disallow :)

https://github.com/slevomat/coding-standard#slevomatcodingstandardcontrolstructuresassignmentincondition

But this could be another discussion because we do that a lot (attribution inside control statement)

**Review by diegofelix — APPROVED** [review]

**Review by brnbp — APPROVED** [review]

**Review by ejosess — APPROVED** [review]

**Comment by gmsantos** [comment]
@carusogabriel I think we should wait for everyone review before merge. The idea is to discuss cons/pros and after that we decide what to do.

**Comment by gmsantos** [comment]
@leroy-merlin-br/boitata if there are anymore objections, Let's merge this

**Comment by carusogabriel** [comment]
@gmsantos Let’s just wait for us to add tests and adopt (or not) SEMVER :)

**Review by djonasm — APPROVED** [review]

**Comment by carusogabriel** [comment]
Depends on #4

File: tests/input/yoda-compairson.php
Comment (ravanscafi) [inline tests/input/yoda-compairson.php]: you forgot to fix the typo on filename

File: tests/expected/yoda-compairson.php
Comment (ravanscafi) [inline tests/expected/yoda-compairson.php]: you forgot to fix the typo on filename

**Review by ravanscafi — CHANGES_REQUESTED** [review]

File: tests/expected/yoda-compairson.php
Comment (gmsantos) [reply tests/expected/yoda-compairson.php]: Codesniffer should fix these kind of typos too :smile:

**Review by ravanscafi — APPROVED** [review]

**Review by djonasm — CHANGES_REQUESTED** [review]
let's allow both ways

File: lib/LeroyMerlin/ruleset.xml
Comment (diegofelix) [inline lib/LeroyMerlin/ruleset.xml:134]: What about just remove this line and dont check this rule at all? its really very rare the case where this may cause a problem and, for me, is just a standard.

**Review by diegofelix — CHANGES_REQUESTED** [review]

File: lib/LeroyMerlin/ruleset.xml
Comment (gmsantos) [reply lib/LeroyMerlin/ruleset.xml:134]: What kind of problem, could you explain better?

Comment (diegofelix) [reply lib/LeroyMerlin/ruleset.xml:134]: A problem when you trying to compare and end up assign a value to a variable.
e.g `if ($count = 2) { `...

But the problem is not the point, the question is: We really need to check if its yoda or not? instead of force the opposite of yoda, just dont check at all.

Comment (gmsantos) [reply lib/LeroyMerlin/ruleset.xml:134]: > We really need to check if its yoda or not?

Yes, we could. It's one of the initial proposals. I'll make clear the voting options and what we voted so far.

**Comment by gmsantos** [comment]
The current options are:

Disallow: @carusogabriel @ravanscafi @ernandojs @gmsantos 
Enforce: -
Neither: @djonasm @diegofelix @ravanscafi @aleferreiranogueira

**Comment by gmsantos** [comment]
@aleferreiranogueira @ernjs @carusogabriel is that what you meant in the review?

**Comment by gmsantos** [comment]
ping @leroy-merlin-br/boitata-backend

**Comment by aleferreiranogueira** [comment]
Actually, I share @diegofelix's and @djonasm's opinions, allowing yoda but not enforcing it. By enforcing it I thing we're trying to protect ourselves from something that do not happen very often and even if it does our unit tests should catch cases where a attribution is being used instead of a comparison.

**Review by alemferreira — CHANGES_REQUESTED** [review]
We should allow both ways

**Pull request closed without merging**
